File: declined-order-still-holds-stock/node/restock-declined-orders.js
/**
 * Restock BigCommerce Declined orders whose stock was never returned.
 *
 * BigCommerce debits inventory_level at order creation, and only returns it when
 * the order reaches a status your Inventory Settings map to "return stock",
 * typically Cancelled or Refunded. Declined (status_id 6) is often not covered,
 * so the debited stock sits withheld from real buyers. This lists recently
 * Declined orders, confirms with GET /v2/orders/{id}/transactions that nothing
 * was actually approved or captured, and returns each line item's quantity with
 * POST /v3/inventory/adjustments/relative. Orders with real money behind them
 * are flagged for a human instead of auto-restocked. Guarded by DRY_RUN. Safe
 * to run again and again.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/declined-order-still-holds-stock/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const BASE_URL = `https://api.bigcommerce.com/stores/${STORE_HASH}/`;
const LOOKBACK_DAYS = Number(process.env.RESTOCK_LOOKBACK_DAYS || 3);
const LOCATION_ID = Number(process.env.LOCATION_ID || 1);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const DECLINED_STATUS_ID = 6;
const CHARGED_STATUSES = new Set(["approved", "captured"]);
const ADJUSTED_NOTE = "declined-order-restocked-by-script";

function headers() {
  return {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    Accept: "application/json",
  };
}

/**
 * Pure decision function. No network calls.
 *
 * order: { status_id, products: [{ variant_id, sku, quantity }], already_adjusted }
 * transactions: [{ status }]
 *
 * Returns { action: "restock" | "flag" | "skip", items: [{ variant_id, qty }] }.
 *
 * Only Declined orders (status_id === 6) are candidates. An order already marked
 * adjusted is skipped so a second run never double-restocks it. If any
 * transaction shows an approved or captured status, money moved despite the
 * Declined status, so the order is flagged for a human instead of touched.
 * Otherwise every line item's quantity is queued to be added back.
 */
export function decideRestock(order, transactions) {
  if (order.status_id !== DECLINED_STATUS_ID) return { action: "skip", items: [] };
  if (order.already_adjusted) return { action: "skip", items: [] };
  if (transactions.some((t) => CHARGED_STATUSES.has(t.status))) return { action: "flag", items: [] };
  const items = order.products.map((p) => ({ variant_id: p.variant_id, qty: p.quantity }));
  return { action: "restock", items };
}

async function bcGet(path, params = {}) {
  const url = new URL(BASE_URL + path);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: headers() });
  if (res.status === 204) return [];
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const text = await res.text();
  return text ? JSON.parse(text) : [];
}

async function bcPost(path, body) {
  const res = await fetch(BASE_URL + path, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const text = await res.text();
  return text ? JSON.parse(text) : null;
}

function rfc1123(date) {
  return date.toUTCString().replace("GMT", "+0000");
}

async function* declinedOrders() {
  const since = rfc1123(new Date(Date.now() - LOOKBACK_DAYS * 86400000));
  let page = 1;
  const limit = 50;
  while (true) {
    const orders = await bcGet("v2/orders", {
      status_id: DECLINED_STATUS_ID,
      min_date_modified: since,
      limit,
      page,
    });
    if (!orders || !orders.length) return;
    for (const order of orders) yield order;
    if (orders.length < limit) return;
    page += 1;
  }
}

async function getOrderProducts(orderId) {
  const products = (await bcGet(`v2/orders/${orderId}/products`)) || [];
  return products.map((p) => ({ variant_id: p.variant_id, sku: p.sku, quantity: p.quantity }));
}

async function getOrderTransactions(orderId) {
  return (await bcGet(`v2/orders/${orderId}/transactions`)) || [];
}

async function getOrderNotes(orderId) {
  return (await bcGet(`v2/orders/${orderId}/notes`)) || [];
}

async function isAlreadyAdjusted(orderId) {
  const notes = await getOrderNotes(orderId);
  return notes.some((n) => (n.note || "").includes(ADJUSTED_NOTE));
}

async function restockItems(items, locationId) {
  const payload = {
    reason: "Declined order restock reconciliation",
    location_id: locationId,
    items: items.map((i) => ({ variant_id: i.variant_id, quantity: i.qty })),
  };
  return bcPost("v3/inventory/adjustments/relative", payload);
}

async function markAdjusted(orderId) {
  return bcPost(`v2/orders/${orderId}/notes`, { note: ADJUSTED_NOTE });
}

export async function run() {
  let restocked = 0;
  let flagged = 0;
  for await (const order of declinedOrders()) {
    const orderId = order.id;
    const fullOrder = {
      status_id: order.status_id ?? DECLINED_STATUS_ID,
      products: await getOrderProducts(orderId),
      already_adjusted: await isAlreadyAdjusted(orderId),
    };
    const transactions = await getOrderTransactions(orderId);
    const decision = decideRestock(fullOrder, transactions);

    if (decision.action === "skip") continue;

    if (decision.action === "flag") {
      console.warn(`Order ${orderId} Declined but has approved/captured transactions. Flagged for manual review.`);
      flagged++;
      continue;
    }

    console.log(
      `Order ${orderId} eligible to restock ${decision.items.length} item(s). ${DRY_RUN ? "would restock" : "restocking"}`
    );
    if (!DRY_RUN) {
      await restockItems(decision.items, LOCATION_ID);
      await markAdjusted(orderId);
    }
    restocked++;
  }
  console.log(
    `Done. ${restocked} order(s) ${DRY_RUN ? "to restock" : "restocked"}, ${flagged} order(s) flagged for review.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: declined-order-still-holds-stock/python/restock_declined_orders.py
"""Restock BigCommerce Declined orders whose stock was never returned.

BigCommerce debits inventory_level at order creation, and only returns it when
the order reaches a status your Inventory Settings map to "return stock",
typically Cancelled or Refunded. Declined (status_id 6) is often not covered,
so the debited stock sits withheld from real buyers. This lists recently
Declined orders, confirms with GET /v2/orders/{id}/transactions that nothing
was actually approved or captured, and returns each line item's quantity with
POST /v3/inventory/adjustments/relative. Orders with real money behind them
are flagged for a human instead of auto-restocked. Guarded by DRY_RUN. Safe to
run again and again.
"""
import os
import logging
from datetime import datetime, timedelta, timezone

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("restock_declined_orders")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
ACCESS_TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
BASE_URL = f"https://api.bigcommerce.com/stores/{STORE_HASH}/"
LOOKBACK_DAYS = int(os.environ.get("RESTOCK_LOOKBACK_DAYS", "3"))
LOCATION_ID = int(os.environ.get("LOCATION_ID", "1"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

DECLINED_STATUS_ID = 6
CHARGED_STATUSES = {"approved", "captured"}
ADJUSTED_NOTE = "declined-order-restocked-by-script"


def _headers():
    return {
        "X-Auth-Token": ACCESS_TOKEN,
        "Content-Type": "application/json",
        "Accept": "application/json",
    }


def decide_restock(order: dict, transactions: list) -> dict:
    """Pure decision function. No network calls.

    order: {"status_id": int, "products": list[{"variant_id", "sku", "quantity"}],
            "already_adjusted": bool}
    transactions: list[{"status": str}]

    Returns {"action": "restock" | "flag" | "skip", "items": list[{"variant_id", "qty"}]}.

    Only Declined orders (status_id == 6) are candidates. An order already marked
    adjusted is skipped so a second run never double-restocks it. If any
    transaction shows an approved or captured status, money moved despite the
    Declined status, so the order is flagged for a human instead of touched.
    Otherwise every line item's quantity is queued to be added back.
    """
    if order["status_id"] != DECLINED_STATUS_ID:
        return {"action": "skip", "items": []}
    if order.get("already_adjusted"):
        return {"action": "skip", "items": []}
    if any(t.get("status") in CHARGED_STATUSES for t in transactions):
        return {"action": "flag", "items": []}
    items = [{"variant_id": p["variant_id"], "qty": p["quantity"]} for p in order["products"]]
    return {"action": "restock", "items": items}


def declined_orders():
    """Yield candidate orders with status_id 6 modified within the lookback window."""
    since = (datetime.now(timezone.utc) - timedelta(days=LOOKBACK_DAYS)).strftime(
        "%a, %d %b %Y %H:%M:%S +0000"
    )
    page = 1
    limit = 50
    while True:
        r = requests.get(
            BASE_URL + "v2/orders",
            headers=_headers(),
            params={
                "status_id": DECLINED_STATUS_ID,
                "min_date_modified": since,
                "limit": limit,
                "page": page,
            },
            timeout=30,
        )
        if r.status_code == 204:
            return
        r.raise_for_status()
        orders = r.json()
        if not orders:
            return
        for order in orders:
            yield order
        if len(orders) < limit:
            return
        page += 1


def get_order_products(order_id):
    r = requests.get(BASE_URL + f"v2/orders/{order_id}/products", headers=_headers(), timeout=30)
    if r.status_code == 204:
        return []
    r.raise_for_status()
    products = r.json() or []
    return [{"variant_id": p["variant_id"], "sku": p.get("sku"), "quantity": p["quantity"]} for p in products]


def get_order_transactions(order_id):
    r = requests.get(BASE_URL + f"v2/orders/{order_id}/transactions", headers=_headers(), timeout=30)
    if r.status_code == 204:
        return []
    r.raise_for_status()
    return r.json() or []


def get_order_notes(order_id):
    r = requests.get(BASE_URL + f"v2/orders/{order_id}/notes", headers=_headers(), timeout=30)
    if r.status_code == 204:
        return []
    r.raise_for_status()
    return r.json() or []


def is_already_adjusted(order_id):
    return any(ADJUSTED_NOTE in (n.get("note") or "") for n in get_order_notes(order_id))


def restock_items(items, location_id):
    payload = {
        "reason": "Declined order restock reconciliation",
        "location_id": location_id,
        "items": [{"variant_id": i["variant_id"], "quantity": i["qty"]} for i in items],
    }
    r = requests.post(
        BASE_URL + "v3/inventory/adjustments/relative",
        headers=_headers(),
        json=payload,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def mark_adjusted(order_id):
    r = requests.post(
        BASE_URL + f"v2/orders/{order_id}/notes",
        headers=_headers(),
        json={"note": ADJUSTED_NOTE},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    restocked = 0
    flagged = 0
    for order in declined_orders():
        order_id = order["id"]
        full_order = {
            "status_id": order.get("status_id", DECLINED_STATUS_ID),
            "products": get_order_products(order_id),
            "already_adjusted": is_already_adjusted(order_id),
        }
        transactions = get_order_transactions(order_id)
        decision = decide_restock(full_order, transactions)

        if decision["action"] == "skip":
            continue

        if decision["action"] == "flag":
            log.warning(
                "Order %s Declined but has approved/captured transactions. Flagged for manual review.",
                order_id,
            )
            flagged += 1
            continue

        log.info(
            "Order %s eligible to restock %d item(s). %s",
            order_id,
            len(decision["items"]),
            "would restock" if DRY_RUN else "restocking",
        )
        if not DRY_RUN:
            restock_items(decision["items"], LOCATION_ID)
            mark_adjusted(order_id)
        restocked += 1

    log.info(
        "Done. %d order(s) %s, %d order(s) flagged for review.",
        restocked,
        "to restock" if DRY_RUN else "restocked",
        flagged,
    )


if __name__ == "__main__":
    run()
